std/experimental/ndslice/algorithm.d

```diff
@@ -0,0 +1,1484 @@
+$(T6 ndCmp, No, No, `2`, `int`, No)
+)
+
+All operators are suitable to change tensors using `ref` argument qualification in a function declaration.
+
+$(H3 Lockstep Iteration)
+
+$(REF_ALTTEXT assumeSameStructure, assumeSameStructure, std,experimental,ndslice,slice)
+can be used as multidimensional `zip` analog if tensors have the same structure (shape and strides).
+`assumeSameStructure` allows to mutate elements of zipped tensors, which is not possible with common
+$(REF zip, std,range).
+
+Also tensors zipped with `assumeSameStructure` uses single set of lengths and strides.
+Thus, `assumeSameStructure` may significantly optimize iteration.
+
+If tensors have different strides, then most of existing operators in this module still
+can be used as they accept a set of tensors instead of single one.
+
+Note:
+    $(SUBREF iteration, transposed) and
+    $(SUBREF selection, pack) can be used to specify dimensions.
+
+License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors:   Ilya Yaroshenko
+
+Source:    $(PHOBOSSRC std/_experimental/_ndslice/_algorithm.d)
+
+Macros:
+SUBREF = $(REF_ALTTEXT $(TT $2), $2, std,experimental, ndslice, $1)$(NBSP)
+T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
+T6=$(TR $(TDNW $(LREF $1)) $(TD $2) $(TD $3) $(TD $4) $(TD $5) $(TD $6))
+T7=$(TR $(TDNW $(LREF $1)) $(TD $2) $(TD $3) $(TD $4) $(TD $5) $(TD $6), $(TD $7))
+*/
+module std.experimental.ndslice.algorithm;
+
+import std.traits;
+import std.meta;
+import std.typecons : Flag, Yes, No;
+
+import std.experimental.ndslice.internal;
+import std.experimental.ndslice.slice;
+
+private template TensorFronts(size_t length)
+{
+    static if (length)
+    {
+        enum i = length - 1;
+        enum TensorFronts = TensorFronts!(length - 1) ~ "tensors[" ~ i.stringof ~ "].front, ";
+    }
+    else
+    {
+        enum TensorFronts = "";
+    }
+}
+
+private void checkShapesMatch(bool seed, Args...)(auto ref Args tensors)
+{
+    enum msg = seed ?
+        "all arguments except the first (seed) must be tensors" :
+        "all arguments must be tensors"
+        ~ tailErrorMessage!();
+    enum msgShape = "all tensors must have the same shape"  ~ tailErrorMessage!();
+    foreach (i, Arg; Args)
+    {
+        static assert (is(Arg == Slice!(N, Range), size_t N, Range), msg);
+        static if (i)
+        {
+            static assert (tensors[i].N == tensors[0].N, msgShape);
+            assert(tensors[i].shape == tensors[0].shape, msgShape);
+        }
+    }
+}
+
+private template naryFun(bool hasSeed, size_t argCount, alias fun)
+{
+    static if (argCount + hasSeed == 1)
+    {
+        import std.functional : unaryFun;
+        alias naryFun = unaryFun!fun;
+    }
+    else
+    static if (argCount + hasSeed == 2)
+    {
+        import std.functional : binaryFun;
+        alias naryFun = binaryFun!fun;
+    }
+    else
+    {
+        alias naryFun = fun;
+    }
+}
+
+/++
+Implements the homonym function (also known as `transform`) present
```

Comment:
This sentence seems overly complex, 

"Implements `map` for `Slice` (also known as `transform`) that's present in many functional languages."

sounds more natural.